src/task/task_graph.rs

```diff
@@ -59,13 +77,28 @@ impl<'p> TaskNode<'p> {
 
 /// A [`TaskGraph`] is a graph of tasks that defines the relationships between different executable
 /// tasks.
+#[derive(Debug)]
 pub struct TaskGraph<'p> {
     /// The project that this graph references
     project: &'p Project,
 
     /// The tasks in the graph
     nodes: Vec<TaskNode<'p>>,
 }
+impl fmt::Display for TaskGraph<'_> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(
+            f,
+            "TaskGraph: number of nodes: {}, nodes: {}",
+            self.nodes.len(),
+            self.nodes
+                .iter()
+                .map(|node| node.to_string())
+                .collect::<Vec<String>>()
+                .join("\n")
```

Comment:
Use `format` instead of `join` to not have to collect into a `Vec<String>` first.

```
self.nodes.iter().format('\n')
```